main/lsp/LSPTypechecker.cc

```diff
@@ -379,11 +383,19 @@ LSPTypechecker::SlowPathResult LSPTypechecker::runSlowPath(LSPFileUpdates &updat
                 case SlowPathMode::Cancelable: {
                     timeit.emplace(this->config->logger, "slow_path_reindex");
 
-                    // Determine which files we need to copy into the open files cache (indexedFinalGS).
+                    // Determine which files we need to copy into the open files cache (indexedFinalGS), and update the
+                    // file table to point to the updated files.
                     if (!updates.updatedFiles.empty()) {
+                        core::UnfreezeFileTable updateFileTable{*this->gs};
+
                         for (auto &file : updates.updatedFiles) {
                             auto fref = this->gs->findFileByPath(file->path());
-                            ENFORCE(fref.exists());
+                            if (!fref.exists()) {
+                                fref = this->gs->enterFile(std::move(file));
+                            } else {
+                                this->gs->replaceFile(fref, std::move(file));
+                            }
```

Comment:
Since we can't rely on getting an up-to-date copy of the indexer's file table anymore, we have to go back to applying these file updates during the slow path. As we apply them in the fast path as well, perhaps we can extract out a convenience helper.